build_runner_core/lib/src/asset_graph/graph.dart

```diff
@@ -374,17 +374,20 @@ class AssetGraph {
   bool _actionMatches(BuildAction action, AssetId input) {
     if (input.package != action.package) return false;
     if (!action.generateFor.matches(input)) return false;
-    Iterable<String> inputExtensions;
+
     if (action is InBuildPhase) {
-      inputExtensions = action.builder.buildExtensions.keys;
+      if (expectedOutputs(action.builder, input).isEmpty) return false;
```

Comment:
We should consider separating out a different function that just checks for matching input extensions, and doesn't also build up the outputs.